src/main/java/store/bookshop/service/order/OrderServiceImpl.java

```diff
@@ -0,0 +1,107 @@
+package store.bookshop.service.order;
+
+import jakarta.transaction.Transactional;
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+import store.bookshop.dto.order.CreateOrderRequestDto;
+import store.bookshop.dto.order.OrderDto;
+import store.bookshop.dto.order.OrderItemDto;
+import store.bookshop.dto.order.UpdateOrderRequestDto;
+import store.bookshop.exception.EntityNotFoundException;
+import store.bookshop.mapper.OrderItemMapper;
+import store.bookshop.mapper.OrderMapper;
+import store.bookshop.model.Book;
+import store.bookshop.model.Order;
+import store.bookshop.model.OrderItem;
+import store.bookshop.repository.BookRepository;
+import store.bookshop.repository.OrderItemRepository;
+import store.bookshop.repository.OrderRepository;
+import store.bookshop.repository.UserRepository;
+
+@Service
+@RequiredArgsConstructor
+public class OrderServiceImpl implements OrderService {
+    private final OrderRepository orderRepository;
+    private final UserRepository userRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final OrderMapper orderMapper;
+    private final OrderItemMapper orderItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public List<OrderDto> findAll(Long userId) {
+        return orderMapper.toDtoList(orderRepository.findByUserId(userId));
+    }
+
+    @Override
+    @Transactional
+    public OrderDto createOrder(Long userId, CreateOrderRequestDto requestDto) {
+        Order newOrder = new Order();
+
+        List<OrderItem> orderItems = orderItemMapper.toModelList(requestDto.getOrderItems());
+        BigDecimal total = new BigDecimal(0);
+        for (OrderItem orderItem : orderItems) {
+            orderItem.setOrder(newOrder);
+            Book book = bookRepository.findById(orderItem.getBook().getId()).orElseThrow(
+                    () -> new EntityNotFoundException("Can't find book by id: "
+                            + orderItem.getBook().getId()));
+            BigDecimal price = book.getPrice();
+            BigDecimal bookPrices = price.multiply(BigDecimal.valueOf(orderItem.getQuantity()));
+            total = total.add(bookPrices);
+            orderItem.setPrice(price);
+        }
+
+        newOrder.setTotal(total);
+        newOrder.setOrderDate(LocalDateTime.now());
+        newOrder.setUser(userRepository.getReferenceById(userId));
+        newOrder.setStatus(Order.Status.COMPLETED);
+        newOrder.setShippingAddress(requestDto.getShippingAddress());
+        Order save = orderRepository.save(newOrder);
+
+        for (OrderItem orderItem : orderItems) {
+            orderItemRepository.save(orderItem);
+        }
```

Comment:
orderItems must be already saved (cascade)
add them to order before `orderRepository.save(newOrder);`

Reply:
HHH000437: Attempting to save one or more entities that have a non-nullable association with an unsaved transient entity. The unsaved transient entity must be saved in an operation prior to saving these dependent entities.